homeassistant/components/volumio/media_player.py

```diff
@@ -272,3 +285,15 @@ async def async_get_browse_image(
         cached_url = self.thumbnail_cache.get(media_content_id)
         image_url = self._volumio.canonic_url(cached_url)
         return await self._async_fetch_image(image_url)
+
+    async def _play_with_mpd(self, media_id):
+        """Play the media_id (url) with mpd because Volumio api does not support it."""
+        await self._volumio.stop()
+        await self._volumio.clear_playlist()
+        client = MPDClient()
```

Comment:
The code owner should preferably approve this before we add it. It's unconventional to use two different interfaces and libraries for the same service.

Reply:
I have to admit I'm not a big fan of this. Volumio has a public API, and I believe that MPD is an implementation detail. Do we know the consequences of calling MPD directly? Is this guaranteed to always be around?

At the very least, I would say this should be encapsulated within PyVolumio, but I think a better approach is to try to add support in Volumio directly - it's also an open source project, and they accept PRs.

Reply:
Another thought - if we have the URL, can't we call the Volumio API to get the item, and then use that to play?

Reply:
I would definitely prefer to do it with the volumio api. I have been trying to find a way to do so, but as far as I know neigther REST nor websocket/socketio api offer a way to do it. I have asked [here](https://community.volumio.org/t/playing-remote-music-content/50781) but did not get an answer yet. I would not call mpd an implementation detail. It is the core to play all the music except for spotify as you can see [here](https://volumio.github.io/docs/Development_How_To/System_Architecture.html).
Moving the code to pyvolumio might be a good idea.

Reply:
The fact that it has an MPD emulation interface is reassuring, but this still doesn't feel right. My first suggestion would be to try to contribute to Volumio API with the missing functionality, and only if that doesn't work for some reason, move the code to PyVolumio.

Reply:
It is not an emulation interface. It is MPD that is run by volumio.
Nevertheless it would be good if volumio would be able to play remote content, so I will keep on trying with a feature request there. Unfortunately I am not a nodejs coder so I cannot contribute the functionality and it not just an API patch (I think).

Reply:
> It is not an emulation interface. It is MPD that is run by volumio.

The link you sent shows an MPD Emulation interface, and that's more important than running on MPD (because it means the interface is more likely to be stay around and be supported).

Reply:
@OnFreund I should have looked closer before my last answer :-). Of course you are right, there is a mpd emulation interface.
With the link above I was trying to point to the 'Music Controllers' and 'Music Services' section.

Reply:
Anyway, knowing the emulation exists makes me comfortable adding this to PyVolumio. My preference is still for this to be supported by Volumio directly, though.